### PR TITLE
add ncps support

### PR DESCRIPTION
**Pull request by amao9098**
Adding ncps support to npt-linux

### CODE REVIEW

File: ncps/ncps.TESTGEN.ps1 [hunk 1]
```diff
@@ -0,0 +1,209 @@
+Param(
+    [Parameter(Mandatory=$false)] [String] $Config = "Default",
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:2]: All of these should be validated

File: ncps/ncps.TESTGEN.ps1 [hunk 2]
```diff
@@ -0,0 +1,209 @@
+Param(
+    [Parameter(Mandatory=$false)] [String] $Config = "Default",
+    [Parameter(Mandatory=$true)]  [String] $DestIp,
+    [Parameter(Mandatory=$true)]  [String] $SrcIp,
+    [Parameter(Mandatory=$true)]  [ValidateScript({Test-Path $_ -PathType Container})] [String] $OutDir = "",
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:5]: remove the default value ( it is invalid if used)

File: ncps/ncps.TESTGEN.ps1 [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    Write-Output " Inputs:"
+    Write-Output "  -Config     = $Config"
+    Write-Output "  -DestIp     = $DestIp"
+    Write-Output "  -SrcIp      = $SrcIp"
+    Write-Output "  -OutDir     = $OutDir"
+    Write-Output "============================================"
+} # input_display()
+
+function banner {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:30]: remove

File: ncps/ncps.TESTGEN.ps1 [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    Param(
+        [parameter(Mandatory=$true)] [String] $Msg
+    )
+    Write-Output "`n==========================================================================="
+    Write-Output "| $Msg"
+    Write-Output "==========================================================================="
+} # banner()
+
+function test_client {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:40]: Remove

File: ncps/ncps.TESTGEN.ps1 [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    Param(
+        [Parameter(Mandatory=$true)] [String] $OutDir,
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:42]: validate your params

File: ncps/ncps.TESTGEN.ps1 [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    $thread_params = "-r $Threads $g_DestIp,$($g_Config.Port)"
+    [String] $cmd = "ncps.exe -s $thread_params -wt $($g_Config.Warmup) -t $($g_Config.Runtime) -o $out.txt $($g_Config.Options)"
+    Write-Output $cmd | Out-File -Encoding ascii -Append "$out.txt"
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_log
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_logRecv  
+    Write-Output   $cmd 
+} # test_send()
+
+function test_iterations {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:82]: Remove

File: ncps/ncps.TESTGEN.ps1 [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    Param(
+        [Parameter(Mandatory=$true)] [String] $OutDir,
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:84]: validate your params

File: ncps/ncps.TESTGEN.ps1 [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    for ($i=0; $i -lt $g_Config.Iterations; $i++) {
+        test_client -OutDir $OutDir -Filename "$Filename.iter$i" -Threads $Threads `
+                    -ConnectionsPerThread $ConnectionsPerThread -ConnectionDurationMS $ConnectionDurationMS `
+                    -DataTransferMode $DataTransferMode -MaxPendingRequests $MaxPendingRequests -SendDir $SendDir
+        test_server -OutDir $OutDir -Filename "$Filename.iter$i" -Threads $Threads -SendDir $SendDir
+    }
+} # test_iterations()
+
+function test_ncps {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:104]: remove

File: ncps/ncps.TESTGEN.ps1 [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    Param(
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:105]: validate params

File: PERFTEST.PS1 [hunk 10]
```diff
@@ -8,8 +8,8 @@ Param(
     [parameter(Mandatory=$true)]  [String] $OutDir = "",
```
Comment (DHarper89936) [inline PERFTEST.PS1:8]: validate params

File: PERFTEST.PS1 [hunk 11]
```diff
@@ -99,7 +99,8 @@ function test_main {
         [parameter(Mandatory=$true)]  [String] $OutDir = "",
```
Comment (DHarper89936) [inline PERFTEST.PS1:99]: validate params

File: PERFTEST.PS1 [hunk 12]
```diff
@@ -99,7 +99,8 @@ function test_main {
...
         [parameter(Mandatory=$true)]  [string] $DestUserName,
         [parameter(Mandatory=$true)]  [string] $SrcUserName,
-        [parameter(Mandatory=$true)]  [string] $TestUserName
+        [parameter(Mandatory=$true)]  [string] $TestUserName,
+        [parameter(Mandatory=$false)] [Array] $ToolList = @("ntttcp", "lagscope", "ncps")
```
Comment (DHarper89936) [inline PERFTEST.PS1:103]: nit: use single quotes

File: PERFTEST.PS1 [hunk 13]
```diff
@@ -8,8 +8,8 @@ Param(
     [parameter(Mandatory=$true)]  [String] $OutDir = "",
     [parameter(Mandatory=$true)]  [string] $DestUserName,
     [parameter(Mandatory=$true)]  [string] $SrcUserName,
-    [parameter(Mandatory=$true)]  [string] $TestUserName
-
+    [parameter(Mandatory=$true)]  [string] $TestUserName,
+    [parameter(Mandatory=$false)] [Array] $ToolList = @("ntttcp", "lagscope", "ncps")
```
Comment (DHarper89936) [inline PERFTEST.PS1:12]: nit: use single quotes

File: PERFTEST.PS1 [hunk 14]
```diff
@@ -125,7 +126,13 @@ function test_main {
     env_create  -OutDir $workDir
 
     # lagscope does not have detail 
-    & "$PSScriptRoot/lagscope/lagscope.TESTGEN.ps1"   -DestIp $g_DestIp -SrcIp $g_SrcIp -OutDir $workDir -DestDir $DestDir -SrcDir $SrcDir -Config $g_config
-    & "$PSScriptRoot/ntttcp/ntttcp.TESTGEN.ps1" -DestIp $g_DestIp -SrcIp $g_SrcIp -OutDir $workDir -Config $g_config -DestDir $DestDir -SrcDir $SrcDir
-
+    if ("lagscope" -in $ToolList) {
```
Comment (DHarper89936) [inline PERFTEST.PS1:129]: nit: single quotes for constant strings

File: ncps/ncps.TESTGEN.ps1 [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+$scriptName = $MyInvocation.MyCommand.Name
+
+#===============================================
+# Internal Functions
+#===============================================
+
+function input_display {
+    $g_path = Get-Location
+
+    Write-Output "============================================"
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:19]: nit: single quotes for constant strings ( applies to lines 21& 26 too

File: ncps/ncps.TESTGEN.ps1 [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+    Write-Output "  -DestIp     = $DestIp"
+    Write-Output "  -SrcIp      = $SrcIp"
+    Write-Output "  -OutDir     = $OutDir"
+    Write-Output "============================================"
+} # input_display()
+
+function banner {
+    [CmdletBinding()]
+    Param(
+        [parameter(Mandatory=$true)] [String] $Msg
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:32]: Validate parameter ( [ValidateScript({-Not [String]::IsNullOrWhiteSpace($_)})] )

File: ncps/ncps.TESTGEN.ps1 [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+    $thread_params = "-r $Threads $g_SrcIp,$($g_Config.Port),$g_DestIp,$($g_Config.Port),$ConnectionsPerThread,$ConnectionsPerThread,$ConnectionDurationMS,$DataTransferMode"
+    [String] $cmd = "ncps.exe -c $thread_params -wt $($g_Config.Warmup) -t $($g_Config.Runtime) -o $out.txt $($g_Config.Options)"
+    Write-Output $cmd | Out-File -Encoding ascii -Append "$out.txt"
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_log
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_logSend
+    Write-Output   $cmd
+} # test_recv()
+
+function test_server {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:63]: Remove

File: ncps/ncps.TESTGEN.ps1 [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+    Param(
+        [Parameter(Mandatory=$true)] [String] $OutDir,
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:65]: validate params

File: ncps/ncps.TESTGEN.ps1 [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+            }
+        }
+    }
+
+} # test_ncps()
+
+function validate_config {
+    $isValid = $true
+
+    $int_vars = @("Iterations", "Port", "Threads", "ConnectionsPerThread", "ConnectionDurationMS", "Warmup", "Runtime")
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:135]: nit: single quotes for constant strings

File: ncps/ncps.TESTGEN.ps1 (hunk 1 above)
Comment (amao9098) [reply ncps/ncps.TESTGEN.ps1:2]: I validate in the main function so we can easily add configs without needing to change the TESTGEN script.

File: ncps/ncps.TESTGEN.ps1 [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+    return $isValid
+} # validate_config()
+
+
+#===============================================
+# External Functions - Main Program
+#===============================================
+function test_main {
+    Param(
+        [Parameter(Mandatory=$false)] [String] $Config = "Default",
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:162]: validate params

File: ncps/ncps.TESTGEN.ps1 [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+        [Parameter(Mandatory=$true)]  [String] $DestIp,
+        [Parameter(Mandatory=$true)]  [String] $SrcIp,
+        [Parameter(Mandatory=$true)]  [ValidateScript({Test-Path $_ -PathType Container})] [String] $OutDir = "",
```
Comment (DHarper89936) [inline ncps/ncps.TESTGEN.ps1:165]: remove default as it is invalid

**Review by DHarper89936 — CHANGES_REQUESTED** [review]
Please address comments

File: PERFTEST.PS1 (hunk 11 above)
Comment (amao9098) [reply PERFTEST.PS1:99]: OutDir does not have to exist as we create Outdir with the UserName due to Linux home directory.

File: PERFTEST.PS1 (hunk 10 above)
Comment (amao9098) [reply PERFTEST.PS1:8]: OutDir does not have to exist as we create Outdir with the UserName due to Linux home directory.

**Review by DHarper89936 — APPROVED** [review]

File: runPerftool.psm1 [hunk 22]
```diff
@@ -468,9 +468,15 @@ Function ProcessToolCommands{
...
+                Copy-Item -Path "$toolpath/rc.local" -Destination "/etc/rc.local" -ToSession $sendPSSession
+                Copy-Item -Path "$toolpath/rc.local" -Destination "/etc/rc.local" -ToSession $recvPSSession
+                Invoke-Command -Session $recvPSSession -ScriptBlock { "`n*   soft    nofile  1048575 `n*   hard    nofile  1048575 " >> /etc/security/limits.conf} 
+                Invoke-Command -Session $sendPSSession -ScriptBlock { "`n*   soft    nofile  1048575 `n*   hard    nofile  1048575 " >> /etc/security/limits.conf} 
+                Invoke-Command -Session $recvPSSession -ScriptBlock { chmod +x /etc/rc.local} 
+                Invoke-Command -Session $sendPSSession -ScriptBlock { chmod +x /etc/rc.local} 
             }
 
-            # Enable execution of tool binaries 
+            # Enable execution of tool binarie s
```
Comment (DHarper89936) [inline runPerftool.psm1:479]: why the extra space?

**Review by DHarper89936 — APPROVED** [review]

**Pull request merged**